Update documentation for Cloudstor in AWS/Azure for 17.06

ddebroy opened the pull request:
### Proposed changes

Updated persistent volume documentation for AWS/Azure. Mention several examples on how to use cloudstor volumes with local/shared backing in AWS.

### Unreleased project version (optional)

17.06

### Related issues (optional)

@friism @kencochrane @FrenchBen

friism commented on docker-for-aws/persistent-data-volumes.md:
```diff
@@ -8,7 +8,17 @@ title: Docker for AWS persistent data volumes
 ## What is Cloudstor?
 
-Cloudstor is a volume plugin managed by Docker. It comes pre-installed and pre-configured in swarms deployed on Docker for AWS. Swarm tasks use a volume created through Cloudstor to mount a persistent data volume that stays attached to the swarm tasks no matter which swarm node they get scheduled or migrated to. Cloudstor relies on shared storage infrastructure provided by AWS to allow swarm task ...
+Cloudstor is a modern volume plugin managed by Docker. It comes pre-installed and pre-configured in Docker Swarms deployed through Docker for AWS. Docker Swarm tasks as well as regular containers can use a volume created through Cloudstor to mount a persistent data volume. In Docker for AWS, Cloudstor can be used with two `backing` options: `local` (that uses EBS) and `shared` (that uses EFS) des ...
```
"managed" > "built"

friism replied:
"through" > "with"

friism replied:
"can be used with" > "has"

friism replied:
@ddebroy is there well-defined nomenclature for "local" and "shared"? I'm worried about "local" in particular since people might confuse with volumes on the instance itself. I think the word Goelzer uses is "relocatable" and Andrea used "attachable" or something.

ddebroy replied:
@friism do you think it will be better/clearer to just expose the options for `backing` as `ebs` and `efs` ? That probably clarifies things a lot better instead of `local`/`shared` and having to explain `local` Cloudstor volume backed by EBS and `shared` Cloudstor volume backed by EFS elsewhere in the docs.

friism replied:
If you think it's responsible I'm very +1 on coming up with _our_ nomenclature that describes the semantics and _not_ the implementation. This makes this stuff way more portable, eg. I can deploy a docker-compose file (referencing a "shared" volume) reliably on AWS and Azure (I know there are a lot of pitfalls).

ddebroy replied:
I agree with your point and that's the reason we have various defaults in place so that commands may work across clouds with minimal changes. 

One additional point to consider is if we do want to implement s3 support down the line, we will probably need something other than `shared` to differentiate EFS from s3 although s3 backing will offer `shared` ish semantics.

One possible set of names could be `direct-attached`/`network-attached` which aligns with the usual storage terminology to distinguish SCSI/ATA from NFS/SMB. For s3 we can probably make something up along the lines of `object-attached`. Thoughts?

friism commented on docker-for-aws/persistent-data-volumes.md:
```diff
@@ -8,7 +8,17 @@ title: Docker for AWS persistent data volumes
+
+## Local Cloudstor volumes
+
+ Workloads running in a Docker Service that require access to low latency/high IOPs persistent storage (e.g. a database engine) can use a `local` Cloudstor volume backed by EBS. It is possible to configure the specific type of EBS volume supported by AWS (e.g. gp2, io1, st1, sc1) that the workload requires during volume creation. Each `local` Cloudstor volume is backed by a single EBS volume. If  ...
```
so in both cases it takes a "few minutes" or "several minutes"

friism replied:
@ddebroy if you think it's a good idea it might be useful to describe how to constrain a service to a particular AZ to avoid having to incur the overhead of shuttling snapshots

ddebroy replied:
@friism the observation was it takes about 2 minutes base for the cross az rescheduling no matter the amount of diffs. Beyond 20 GB of diffs, the snapshot process seems to take greater than 10 minutes and rises for additionals GBs.

friism replied:
cool, I'm not sure that's clear from the current "few minutes" vs. "several minutes" terminology

ddebroy replied:
ok will clarify with the above details.

friism commented on docker-for-aws/persistent-data-volumes.md:
```diff
@@ -8,7 +8,17 @@ title: Docker for AWS persistent data volumes
+
+ Workloads running in a Docker Service that require access to low latency/high IOPs persistent storage (e.g. a database engine) can use a `local` Cloudstor volume backed by EBS. It is possible to configure the specific type of EBS volume supported by AWS (e.g. gp2, io1, st1, sc1) that the workload requires during volume creation. Each `local` Cloudstor volume is backed by a single EBS volume. If  ...
+ 
+ Sharing/mounting the same Cloudstor volume backed by EBS among multiple tasks is not a supported scenario and will lead to data loss. If you need a Cloudstor volume to share data between tasks please read below for EFS backed `shared` volume options. `local` Cloudstor backed by EBS is supported on all AWS regions that support EBS. It is the default "backing" option if EFS support is not selected ...
```
what happens if I specify "shared" in a region without EFS support? Am I handed an EBS-based volume? That's probably not good...

ddebroy replied:
You should get an error if you specify `shared` in a region without EFS support. If you don't specify a backing type, that's when you let cloudstor choose for you - you get EFS if EFS is supported and enabled otherwise EBS.

friism replied:
ah, sorry - I get it know

friism commented on docker-for-aws/persistent-data-volumes.md:
```diff
@@ -22,44 +32,86 @@ f416c95c0dcc        cloudstor:aws               cloud storage plugin for Docker
+### Use a unique volume per task (using EBS):
 
-### Configure IO performance
+A unique `local` Cloudstor volume backed by a specified type of EBS can be created and mounted for each task in a swarm service using a templatized notation with the `docker service create` CLI. Creation of new EBS volumes typically takes a few minutes.  The volume options (besides `backing=local` indicating `local` EBS backed volumes) are:
```
double space

friism replied:
what does the container see while the ebs volume is provisioning? or is the task not scheduled before the volume is ready? (if that's the case, maybe make a note of that)

ddebroy replied:
Task is not scheduled.

friism replied:
great

friism commented on docker-for-aws/persistent-data-volumes.md:
````diff
@@ -22,44 +32,86 @@ f416c95c0dcc        cloudstor:aws               cloud storage plugin for Docker
+```
 
-You can use `docker volume ls` to enumerate all volumes created on a node including those backed by Cloudstor. Note that if a swarm service task starts off in a node and has a Cloudstor volume associated and later gets rescheduled to a different node, `docker volume ls` in the initial node will continue to list the Cloudstor volume that was created for the task that no longer executes on the node ...
+### Use a unique volume per task (using EBS):
````
oh, so there's actually kinda three modes: shared, local (single) and local (for each task). Not  critical, but it might be good to very briefly mention that in the intro blurb at the beginning of the document

ddebroy replied:
sounds good.

friism commented on docker-for-aws/persistent-data-volumes.md:
```diff
@@ -22,44 +32,86 @@ f416c95c0dcc        cloudstor:aws               cloud storage plugin for Docker
+
+### Use a unique volume per task (using EFS with EFS support present and enabled):
+
+It is possible to use the templatized notation to indicate to Docker Swarm that a unique EFS backed volume be created and mounted for each replica/task of a service. Example:
```
maybe mention why you would do this. (I can't come up with a reason off the top of my hat)

friism replied:
oh, to avoid the overhead of snapshot-shuffles?

ddebroy replied:
We specify that users should use EBS only when necessary - specific IOPs/latency requirements that EFS cannot deliver. So if the user has a scenario where the tasks in a service write to say, the same files under the same paths but they don't have stringent IOPs/latency requirements of EBS, they should use this "per-task" EFS option. It also saves on snapshot overhead during reschedules.

friism replied:
I see - just putting it out there, but people might want EBS for this since it's a lot cheaper.

mstanleyjones commented:
Please ping after you have addressed the technical feedback, because I'd like to give this a copy-edit pass.

kencochrane commented:
@ddebroy looks good. You mention it briefly, but we should probably describe how to clean up volumes and snapshots that are left over, and maybe call it out a little more. Because they are created outside of CloudFormation, people will need to manually remove these before they can delete their stack. We should mention that, so that we can point people to the docs when it happens and they email us.

ddebroy commented:
Addressed review comments so far above. I have added a note on how to get the list of EBS created by the stack that one can manually delete. So far, I have found that the stack deletion goes through fine even with EBS volumes from cloudstor lying around but those volumes just never get deleted.

ddebroy commented:
@mstanleyjones this is now ready as far as internal tech reviews are concerned.

mstanleyjones commented:
Thanks, I'm doing copy-editing now. I'll have it finished by GA or very shortly after.

FrenchBen commented on docker-for-aws/persistent-data-volumes.md:
```diff
@@ -22,44 +36,88 @@ f416c95c0dcc        cloudstor:aws               cloud storage plugin for Docker
-docker service create --replicas 5 --name ping2 \
-    --mount type=volume,volume-driver=cloudstor:aws,source={{.Service.Name}}-{{.Task.Slot}}-vol,destination=/mydata \
+docker service create --replicas 5 --name ping3 \
+    --mount type=volume,volume-driver=docker4x/cloudstor:aws-v{{ edition_version }},source={{.Service.Name}}-{{.Task.Slot}}-vol5,destination=/mydata,volume-opt=perfmode=maxio \
```
Since we alias the plugin to prevent upgrade failures, shouldn't this `volume-driver` remain `cloudstor:aws`?

mstanleyjones commented:
@ddebroy PTAL at the copyedit to be sure I didn't change any technical stuff in an incorrect way.

FrenchBen commented on docker-for-aws/persistent-data-volumes.md:
```diff
@@ -8,58 +8,252 @@ title: Docker for AWS persistent data volumes
+$ docker service create \
+  --replicas 5 \
+  --name ping3 \
+  --mount type=volume,volume-driver=docker4x/cloudstor:aws-v{{ edition_version }},source={{.Service.Name}}-{{.Task.Slot}}-vol5,destination=/mydata,volume-opt=perfmode=maxio \
```
Same as before - Since we alias the plugin to prevent upgrade failures, shouldn't this `volume-driver` remain `cloudstor:aws`?

ddebroy commented:
@mstanleyjones  LGTM.

mstanleyjones commented on docker-for-aws/persistent-data-volumes.md:
```diff
@@ -14,8 +14,8 @@ mode tasks and regular Docker containers can use a volume created with
 
-- `relocatable` data volumes are backed by EBS.
-- `shared` data volumes are backed by EFS.
+- `relocatable` data volumes which are backed by EBS.
```
I don't think this is really needed. You took a full sentence and turned it into a fragment.

mstanleyjones commented on docker-for-aws/persistent-data-volumes.md:
````diff
@@ -167,16 +163,14 @@ the following volume options are available:
 ```
 
-This example creates and mounts a distinct Cloudstor volume backed by 25 GB EBS
+The above example creates and mounts a distinct Cloudstor volume backed by 25 GB EBS
````
I don't think this change is necessary.

mstanleyjones commented on docker-for-aws/persistent-data-volumes.md:
````diff
@@ -215,13 +209,11 @@ you already have too many EBS volumes or want to reduce the amount of time it
 takes to transfer volume data across availability zones.
 
 ```bash
-{% raw %}
````
You do need these because of the double curly braces below, which will be interpreted as Liquid.

mstanleyjones replied:
That's correct.

ddebroy commented:
@mstanleyjones I have addressed/undid the changes you requested.
